### DISCUSSION TITLE
Import and Export volumes vs Octopus Mini data

### DISCUSSION THREAD

**Discussion by onetoomany** [BottlecapDave/HomeAssistant-OctopusEnergy#523] [Q&A]
I've just set up my home mini, and I'm trying to understand the data definitions. I've got solar panels etc., so I can be both importing and exporting in any period.
However, I don't have an export contract.

I'm assuming that:
_current_accumulative_consumption = sum of import (as opposed to being net consumption). I assume this is an all-time accumulative?
_current_accumulative_cost = sum of import
_current_demand = if > 0 then Import; if < 0 then Export
_current_consumption = not quite sure what this is - the units and values don't make immediate sense

If I want 
_current_accumulative_export , then I will need to integrate _current_demand when negative

Also, is it possible to get more frequent data updates on _current_consumption than every minute? - as I understand the data is read every 10 seconds.

**Comment by BottlecapDave** [comment]
The sensors are detailed in https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#home-mini-entities. The main thing to note (which I have on my list to update) is that consumption is in `kwh` whereas demand is in `watts`. 

Unfortunately, the consumption data doesn't seem to go negative at any granular level so demand is the only way to indicate that you're exporting data, but this is "live" and so if the API or HA is down for whatever reason, then the accumulation (if done through something like a utility meter) won't be accurate. There doesn't appear to be anything in the API to retrieve live export data - I don't have any solar so I can't probe the API to find out.

> Also, is it possible to get more frequent data updates on _current_consumption than every minute? - as I understand the data is read every 10 seconds.

As per the [docs](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/setup_account.md#refresh-rate-in-minutes)

> The Octopus Energy APIs have a rate limit of 100 calls per hour, which is shared among all calls including through the app. This is usually enough for one meter's data to be retrieved once a minute. However, if you are using other integrations, have more than one meter (e.g. gas) or want the app to not be effected you may want to increase this rate [via the configuration].

Because of this, I don't expose anything lower than 1 minute.

**Reply by theneilkirk** [reply]
Hi

First off, thanks for creating this HA integration @BottlecapDave 

I've just received my Home Mini and also have solar without batteries. I'd love to help get export data into HA, so am willing to share any required info that might help.

Here are two API calls, one for import and one for export for the same time period, when I seem to have been producing more than I've been consuming. For example, Between 12-12:30, my import is 0 and export is 0.251.

I've not come across any negative values on my import...

Import query
`{{url}}/v1/electricity-meter-points/{{import mpan}}/meters/{{import serial}}/consumption/?period_from=2023-11-15T09:00:00&period_to=2023-11-15T14:00:00`

Response
```{
    "count": 11,
    "next": null,
    "previous": null,
    "results": [
        {
            "consumption": 0.207,
            "interval_start": "2023-11-15T14:00:00Z",
            "interval_end": "2023-11-15T14:30:00Z"
        },
        {
            "consumption": 0.025,
            "interval_start": "2023-11-15T13:30:00Z",
            "interval_end": "2023-11-15T14:00:00Z"
        },
        {
            "consumption": 0.002,
            "interval_start": "2023-11-15T13:00:00Z",
            "interval_end": "2023-11-15T13:30:00Z"
        },
        {
            "consumption": 0.153,
            "interval_start": "2023-11-15T12:30:00Z",
            "interval_end": "2023-11-15T13:00:00Z"
        },
        {
            "consumption": 0.0,
            "interval_start": "2023-11-15T12:00:00Z",
            "interval_end": "2023-11-15T12:30:00Z"
        },
        {
            "consumption": 0.001,
            "interval_start": "2023-11-15T11:30:00Z",
            "interval_end": "2023-11-15T12:00:00Z"
        },
        {
            "consumption": 0.0,
            "interval_start": "2023-11-15T11:00:00Z",
            "interval_end": "2023-11-15T11:30:00Z"
        },
        {
            "consumption": 0.026,
            "interval_start": "2023-11-15T10:30:00Z",
            "interval_end": "2023-11-15T11:00:00Z"
        },
        {
            "consumption": 0.157,
            "interval_start": "2023-11-15T10:00:00Z",
            "interval_end": "2023-11-15T10:30:00Z"
        },
        {
            "consumption": 0.23,
            "interval_start": "2023-11-15T09:30:00Z",
            "interval_end": "2023-11-15T10:00:00Z"
        },
        {
            "consumption": 0.026,
            "interval_start": "2023-11-15T09:00:00Z",
            "interval_end": "2023-11-15T09:30:00Z"
        }
    ]
}
```

Export query
`{{url}}/v1/electricity-meter-points/{{export mpan}}/meters/{{export serial}}/consumption/?period_from=2023-11-15T09:00:00&period_to=2023-11-15T14:00:00`

Response
```
{
    "count": 11,
    "next": null,
    "previous": null,
    "results": [
        {
            "consumption": 0.0,
            "interval_start": "2023-11-15T14:00:00Z",
            "interval_end": "2023-11-15T14:30:00Z"
        },
        {
            "consumption": 0.002,
            "interval_start": "2023-11-15T13:30:00Z",
            "interval_end": "2023-11-15T14:00:00Z"
        },
        {
            "consumption": 0.044,
            "interval_start": "2023-11-15T13:00:00Z",
            "interval_end": "2023-11-15T13:30:00Z"
        },
        {
            "consumption": 0.074,
            "interval_start": "2023-11-15T12:30:00Z",
            "interval_end": "2023-11-15T13:00:00Z"
        },
        {
            "consumption": 0.251,
            "interval_start": "2023-11-15T12:00:00Z",
            "interval_end": "2023-11-15T12:30:00Z"
        },
        {
            "consumption": 0.312,
            "interval_start": "2023-11-15T11:30:00Z",
            "interval_end": "2023-11-15T12:00:00Z"
        },
        {
            "consumption": 0.496,
            "interval_start": "2023-11-15T11:00:00Z",
            "interval_end": "2023-11-15T11:30:00Z"
        },
        {
            "consumption": 0.557,
            "interval_start": "2023-11-15T10:30:00Z",
            "interval_end": "2023-11-15T11:00:00Z"
        },
        {
            "consumption": 0.322,
            "interval_start": "2023-11-15T10:00:00Z",
            "interval_end": "2023-11-15T10:30:00Z"
        },
        {
            "consumption": 0.308,
            "interval_start": "2023-11-15T09:30:00Z",
            "interval_end": "2023-11-15T10:00:00Z"
        },
        {
            "consumption": 0.602,
            "interval_start": "2023-11-15T09:00:00Z",
            "interval_end": "2023-11-15T09:30:00Z"
        }
    ]
}
```

**Reply by BottlecapDave** [reply]
The home mini uses different APIs to the ones you've posted, as the ones you've posted are for retrieving data straight from you meter, which is updated at a delay. The home mini data is only available via their graphql endpoints. This [query](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/custom_components/octopus_energy/api_client/__init__.py#L25) is used to get meter information. I use `smartExportElectricityMeter` or `smartImportElectricityMeter` to determine if the meter represents an import or export.

The device id can then be used and passed into this [query](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/custom_components/octopus_energy/api_client/__init__.py#L104) which retrieves the home mini data. If this returns valid live export data, then I can expose this via the integration.

**Reply by theneilkirk** [reply]
Ah, that makes sense. I'm only a novice-level dev and not at all familiar with graphql, so I'll need to find some time to get my head around how to make those calls.

In the meantime, I did note that my HA card showing current demand dropped to a massive negative value this morning while the sun was out. So it seems it would indicate export.
![signal-2023-11-19-090157](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/38278363/4cc9ab30-d6f2-4244-ad50-c33724ae2aa6)

I'll let you know how I get on with the graphql. Alternatively, let me know if you'd be comfortable doing so on my behalf...

**Comment by peterbrown2** [comment]
Has anyone made any progress on this? I would also love to see how much I am exporting (I'm not on an export tariff).
